docs/_docs/02_testing-best-practices/unit-integration-end-to-end-test.md

```diff
@@ -22,3 +22,149 @@ Link to the video at [infoq.com](https://www.infoq.com/presentations/automated-t
 ## Slides
 
 Slides to the video can be found here: [Slides: How to test infrastructure code](https://www.slideshare.net/brikis98/how-to-test-infrastructure-code-automated-testing-for-terraform-kubernetes-docker-packer-and-more){:target="_blank"}.
+
+## Unit test helpers
+
+The function terraform.UnitTest will help unit test a single module, the UnitTest function will do the following things
+
+**_TODO_**: As this is a one day project I will only attempt to do step 3, 4, and 6 today. Time permitting I will add 2
+as well 
+
+1. Copy your terraform code to a temporary folder
+2. Optionally call a setup function you provide
+3. Init and Apply the terraform code
+4. Call a list of validators you provide
+5. Optionally call a tearDown function you provide
+6. Destroy the terraform code
+
+Most of these steps are wrapped into stages which you can skip (see [Iterating locally using test stages]({% link _docs/02_testing-best-practices/iterating-locally-using-test-stages.md %})). This means you can skip individual stages
+by setting env var SKIP_\<stage_name\> to any value.
+
+If you skip the init_apply step then UnitTest will automatically load your previously applied outputs from the working directory.
+
+The full list of stages is:
+
+* setup
+* init_apply
+* validate
+* teardown
+* destroy
+
+Signature of UnitTest:
+
+```go
+// The output of the module after it has been applied, as returned by terraform.OutputAll
+type ModuleOutputs map[string]interface{}
+
+// A validation function to test your terraform code
+type Validator func(t *testing.T, moduleOutputs *terraform.ModuleOutputs, workingDir string)
+
+type UnitTestPlan struct {
+  TerraformOptions *terraform.Options
+  // If you do not provide a working directory it will default to ./.terratest-unit-test/TEST_FUNCTION_NAME/
+  WorkingDirectory string
+  Setup func(t *testing.T, workingDir string)
+  TearDown func(t *testing.T, moduleOutputs *terraform.ModuleOutputs, workingDir string)
+  Validators []Validator
+}
+
+func UnitTest(t *testing.T, testPlan *terraform.UnitTestPlan)
+```
+
+Examples:
+
+This example shows the simplest example with multiple Validators, with the default working directory
+```go
+func TestWebServer(t *testing.T) {
+  terraformOptions := &terraform.Options {
+    // The path to where your Terraform code is located
+    TerraformDir: "../web-server",
+  }
+
+  terrform.UnitTest(&terraform.UnitTestPlan{

```

Comment:
If you need to run `apply` on several modules, how would you compose multiple `UnitTest` calls together? Or is that where integration test comes in?